README.rst
CMS Plugin Google Map
====================

An app for integrating a google map in your cms layout. This is an exact
copy of [django-cms googlemape](https://github.com/divio/django-cms/tree/develop/cms/plugins/googlemap).

The only difference is that all [fields](https://github.com/bitmazk/cmsplugin-googlemap/blob/master/cmsplugin_googlemap/models.py#L25)
except city are now optional.

Prerequisites
-------------

You will need to have the following packages installed:

* Django
* django-cms
* Pillow
* South


Installation
------------

If you want to install the latest stable release from PyPi::

    $ pip install cmsplugin-googlemap

If you feel adventurous and want to install the latest commit from GitHub::

    $ pip install -e git://github.com/bitmazk/cmsplugin-googlemap.git#egg=cmsplugin_googlemap

Add ``cmsplugin_googlemap`` to your ``INSTALLED_APPS``::

    INSTALLED_APPS = (
        ...,
        'cmsplugin_googlemap',
    )

If you were using the googlemap plugin that comes with django-cms you need
uninstall it first, like so::

    ./manage.py cms uninstall plugins GoogleMapPlugin
    psql -U role dbname -h localhost
    DROP TABLE cmsplugin_googlemap;

Run the South migrations::

    ./manage.py migrate cmsplugin_googlemap


Usage
-----

Simply add the plugin to your cms placeholders.

Roadmap
-------

Check the issue tracker on github for milestones and features to come.


CHANGELOG.txt
=== ongoing ===


=== 0.1.5 ===

- removed django-cms from requirements in setup.py, but it is still required

=== 0.1.4 ===

- using small zoom controls to fix display

=== 0.1.3 ===

- removed that ugly white info window

=== 0.1.2 ===

- fixed typo in app name

=== 0.1.1 ===

- Fixed wrong MANIFEST and reinitialized repo

=== 0.1 ===

Initial commit


cmsplugin_googlemap/tests/test_settings.py
"""Settings that need to be set in order to run the tests."""
import os

DEBUG = True
FILER_DEBUG = True

SITE_ID = 1

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.sqlite3",
        "NAME": ":memory:",
    }
}

USE_I18N = True

LANGUAGE_CODE = 'en-us'

LANGUAGES = (
    ('en', 'English'),
    ('de', 'German'),
)

ROOT_URLCONF = 'cmsplugin_googlemap.tests.urls'

STATIC_URL = '/static/'
MEDIA_URL = '/media/'

STATIC_ROOT = os.path.join(os.path.dirname(__file__), '../../static/')
MEDIA_ROOT = os.path.join(os.path.dirname(__file__), '../../media/')

STATICFILES_DIRS = (
    os.path.join(os.path.dirname(__file__), 'test_static'),
)

TEMPLATE_DIRS = (
    os.path.join(os.path.dirname(__file__), '../templates'),
)

COVERAGE_REPORT_HTML_OUTPUT_DIR = os.path.join(
    os.path.dirname(__file__), 'coverage')

COVERAGE_MODULE_EXCLUDES = [
    'tests$', 'settings$', 'urls$', 'locale$',
    'migrations', 'fixtures', 'admin$', 'django_extensions',
]

MIDDLEWARE_CLASSES = [
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'cms.middleware.page.CurrentPageMiddleware',
    'cms.middleware.user.CurrentUserMiddleware',
    'cms.middleware.toolbar.ToolbarMiddleware',
    'simple_translation.middleware.MultilingualGenericsMiddleware',
]

TEMPLATE_CONTEXT_PROCESSORS = (
    'django.contrib.auth.context_processors.auth',
    'django.core.context_processors.i18n',
    'django.core.context_processors.request',
    'django.core.context_processors.media',
    'django.core.context_processors.static',
    'cms.context_processors.media',
    'sekizai.context_processors.sekizai',
)

EXTERNAL_APPS = [
    'django.contrib.admin',
    'django.contrib.admindocs',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.messages',
    'django.contrib.sessions',
    'django.contrib.staticfiles',
    'django.contrib.sitemaps',
    'django.contrib.sites',
    'django_nose',
    'cms',
    'sekizai',
    'menus',
    'mptt',
]

INTERNAL_APPS = [
    'cmsplugin_googlemap',
]


SECRET_KEY = 'this is no real secret key'

INSTALLED_APPS = EXTERNAL_APPS + INTERNAL_APPS

COVERAGE_MODULE_EXCLUDES += EXTERNAL_APPS

CMS_SOFTROOT = True
CMS_PERMISSION = False
CMS_SEO_FIELDS = True
CMS_MENU_TITLE_OVERWRITE = True
CMS_FRONTEND_LANGUAGES = ('en', 'de', )
CMS_TEMPLATES = (
    ('standard.html', 'Standard'),
)


cmsplugin_googlemap/forms.py
"""Forms for the ``cmsplugin_googlemap`` app."""
import re

from django.forms.models import ModelForm
from django.utils.translation import ugettext_lazy as _

from .models import GoogleMap

CSS_WIDTH_RE = re.compile(r'^\d+(?:px|%)$')
CSS_HEIGHT_RE = re.compile(r'^\d+px$')


class GoogleMapForm(ModelForm):
    class Meta:
        model = GoogleMap

    def clean(self):
        cleaned_data = super(GoogleMapForm, self).clean()
        width = cleaned_data.get('width', '')
        height = cleaned_data.get('height', '')
        if width or height:
            if width and not CSS_WIDTH_RE.match(width):
                self._errors['width'] = self.error_class([
                    _(u'Must be a positive integer followed by "px" or "%".')])
            if height and not CSS_HEIGHT_RE.match(height):
                self._errors['height'] = self.error_class([
                    _(u'Must be a positive integer followed by "px".')])
        return cleaned_data


cmsplugin_googlemap/models.py
"""Models for the ``cmsplugin_googlemap`` app."""
from django.db import models
from django.utils.translation import ugettext_lazy as _

from cms.models import CMSPlugin


ZOOM_LEVELS = map(lambda c: (c, str(c)), range(22))


class GoogleMap(CMSPlugin):
    """
    A google maps integration

    """
    title = models.CharField(
        verbose_name=_("map title"),
        max_length=100,
        blank=True,
    )

    address = models.CharField(
        verbose_name=_("address"),
        max_length=150,
        blank=True,
    )

    zipcode = models.CharField(
        verbose_name=_("zip code"),
        max_length=30,
        blank=True,
    )

    city = models.CharField(
        verbose_name=_("city"),
        max_length=100,
    )

    content = models.CharField(
        verbose_name=_("additional content"),
        max_length=255,
        blank=True,
        help_text=_('Displayed under address in the bubble.'),
    )

    zoom = models.PositiveSmallIntegerField(
        verbose_name=_("zoom level"),
        choices=ZOOM_LEVELS,
        default=13,
    )

    lat = models.DecimalField(
        verbose_name=_('latitude'),
        max_digits=10,
        decimal_places=6,
        null=True, blank=True,
        help_text=_('Use latitude & longitude to fine tune the map position.'),
    )

    lng = models.DecimalField(
        verbose_name=_('longitude'),
        max_digits=10,
        decimal_places=6,
        null=True, blank=True,
    )

    route_planer_title = models.CharField(
        verbose_name=_("route planer title"),
        max_length=150,
        blank=True,
        null=True,
        default=_('Calculate your fastest way to here'),
    )

    route_planer = models.BooleanField(
        verbose_name=_("route planer"),
        default=False,
    )

    width = models.CharField(
        verbose_name=_('width'),
        max_length=6,
        default='100%',
        help_text=_('Plugin width (in pixels or percent).')
    )
    height = models.CharField(
        verbose_name=_('height'),
        max_length=6,
        default='400px',
        help_text=_('Plugin height (in pixels).'),
    )

    def __unicode__(self):
        return u"%s (%s, %s %s)" % (self.get_title(), self.address,
                                    self.zipcode, self.city,)

    def get_title(self):
        if self.title is None:
            return _("Map")
        return self.title

    def get_lat_lng(self):
        if self.lat and self.lng:
            return (self.lat, self.lng)


cmsplugin_googlemap/cms_plugins.py
"""CMS Plugins for the ``cmsplugin_googlemap`` app."""
from django.utils.translation import ugettext_lazy as _

from cms.plugin_pool import plugin_pool
from cms.plugin_base import CMSPluginBase

from .models import GoogleMap
from .forms import GoogleMapForm


class GoogleMapPlugin(CMSPluginBase):
    model = GoogleMap
    name = _("Google Map")
    render_template = "cmsplugin_googlemap/googlemap.html"
    admin_preview = False
    form = GoogleMapForm
    fieldsets = (
        (None, {
            'fields': ('title', 'address', ('zipcode', 'city',),
                       'content', 'zoom', ('lat', 'lng'),),
        }),
        (_('Advanced'), {
            'fields': (('route_planer', 'route_planer_title'),
                       ('width', 'height',),),
        }),
    )

    def render(self, context, instance, placeholder):
        context.update({
            'object': instance,
            'placeholder': placeholder,
        })
        return context

plugin_pool.register_plugin(GoogleMapPlugin)


cmsplugin_googlemap/migrations/0001_initial.py
# -*- coding: utf-8 -*-
import datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Adding model 'GoogleMap'
        db.create_table(u'cmsplugin_googlemap', (
            (u'cmsplugin_ptr', self.gf('django.db.models.fields.related.OneToOneField')(to=orm['cms.CMSPlugin'], unique=True, primary_key=True)),
            ('title', self.gf('django.db.models.fields.CharField')(max_length=100, blank=True)),
            ('address', self.gf('django.db.models.fields.CharField')(max_length=150, blank=True)),
            ('zipcode', self.gf('django.db.models.fields.CharField')(max_length=30, blank=True)),
            ('city', self.gf('django.db.models.fields.CharField')(max_length=100)),
            ('content', self.gf('django.db.models.fields.CharField')(max_length=255, blank=True)),
            ('zoom', self.gf('django.db.models.fields.PositiveSmallIntegerField')(default=13)),
            ('lat', self.gf('django.db.models.fields.DecimalField')(null=True, max_digits=10, decimal_places=6, blank=True)),
            ('lng', self.gf('django.db.models.fields.DecimalField')(null=True, max_digits=10, decimal_places=6, blank=True)),
            ('route_planer_title', self.gf('django.db.models.fields.CharField')(default=u'Calculate your fastest way to here', max_length=150, null=True, blank=True)),
            ('route_planer', self.gf('django.db.models.fields.BooleanField')(default=False)),
            ('width', self.gf('django.db.models.fields.CharField')(default='100%', max_length=6)),
            ('height', self.gf('django.db.models.fields.CharField')(default='400px', max_length=6)),
        ))
        db.send_create_signal(u'cmsplugin_googlemap', ['GoogleMap'])


    def backwards(self, orm):
        # Deleting model 'GoogleMap'
        db.delete_table(u'cmsplugin_googlemap')


    models = {
        'cms.cmsplugin': {
            'Meta': {'object_name': 'CMSPlugin'},
            'changed_date': ('django.db.models.fields.DateTimeField', [], {'auto_now': 'True', 'blank': 'True'}),
            'creation_date': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'language': ('django.db.models.fields.CharField', [], {'max_length': '15', 'db_index': 'True'}),
            'level': ('django.db.models.fields.PositiveIntegerField', [], {'db_index': 'True'}),
            'lft': ('django.db.models.fields.PositiveIntegerField', [], {'db_index': 'True'}),
            'parent': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['cms.CMSPlugin']", 'null': 'True', 'blank': 'True'}),
            'placeholder': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['cms.Placeholder']", 'null': 'True'}),
            'plugin_type': ('django.db.models.fields.CharField', [], {'max_length': '50', 'db_index': 'True'}),
            'position': ('django.db.models.fields.PositiveSmallIntegerField', [], {'null': 'True', 'blank': 'True'}),
            'rght': ('django.db.models.fields.PositiveIntegerField', [], {'db_index': 'True'}),
            'tree_id': ('django.db.models.fields.PositiveIntegerField', [], {'db_index': 'True'})
        },
        'cms.placeholder': {
            'Meta': {'object_name': 'Placeholder'},
            'default_width': ('django.db.models.fields.PositiveSmallIntegerField', [], {'null': 'True'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'slot': ('django.db.models.fields.CharField', [], {'max_length': '50', 'db_index': 'True'})
        },
        u'cmsplugin_googlemap.googlemap': {
            'Meta': {'object_name': 'GoogleMap', 'db_table': "u'cmsplugin_googlemap'", '_ormbases': ['cms.CMSPlugin']},
            'address': ('django.db.models.fields.CharField', [], {'max_length': '150', 'blank': 'True'}),
            'city': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            u'cmsplugin_ptr': ('django.db.models.fields.related.OneToOneField', [], {'to': "orm['cms.CMSPlugin']", 'unique': 'True', 'primary_key': 'True'}),
            'content': ('django.db.models.fields.CharField', [], {'max_length': '255', 'blank': 'True'}),
            'height': ('django.db.models.fields.CharField', [], {'default': "'400px'", 'max_length': '6'}),
            'lat': ('django.db.models.fields.DecimalField', [], {'null': 'True', 'max_digits': '10', 'decimal_places': '6', 'blank': 'True'}),
            'lng': ('django.db.models.fields.DecimalField', [], {'null': 'True', 'max_digits': '10', 'decimal_places': '6', 'blank': 'True'}),
            'route_planer': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'route_planer_title': ('django.db.models.fields.CharField', [], {'default': "u'Calculate your fastest way to here'", 'max_length': '150', 'null': 'True', 'blank': 'True'}),
            'title': ('django.db.models.fields.CharField', [], {'max_length': '100', 'blank': 'True'}),
            'width': ('django.db.models.fields.CharField', [], {'default': "'100%'", 'max_length': '6'}),
            'zipcode': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'zoom': ('django.db.models.fields.PositiveSmallIntegerField', [], {'default': '13'})
        }
    }

    complete_apps = ['cmsplugin_googlemap']